### ISSUE TITLE
Live editing

### ISSUE THREAD

**Issue by ikozyris** [ikozyris/yocto#6]
As almost all text editors do, being able to read the file while editing is a must feature.
For this issue to close, the following need to be complete:
- [x] Scroll on text (pager)
- Insertion
    - [x] Screen
    - [x] File
- [x] Files can be saved and edited (without bugs)
- [x] There are no issues with line wrapping
- Features from previous releases get ported
    - [x] Makefile
    - [x] Reading
    - [x] Line numbers
- Special keys work
    - [x] Arrow keys
    - [x] Enter
    - [x] Control *
    - [x] Backspace + Delete
- [x] Unicode (wchar_t)
- [x] Make code modular and separate it in functions and files

**Comment by ikozyris** [comment]
check out the ncurses branch to follow the development

**Comment by ikozyris** [comment]
Problem with modifying text:
New chars need to be inserted into the text at any position without overwriting the previous char which was there (all other char move to the right).

To do this with *O(1)* complexity, an **adjacency list** is needed, however in an adjacency list we cannot say: list[5], the whole line needs to be traversed, *O(n)*.

In the current implementation, which is a **2d array** (adjacency matrix), we can only insert in *O(n)*, which is much, much worse. However, the typical line of a text does not exceed 100 characters, so it not that inefficient to shift ~100 chars after the inserted one forward. In addition, we can access any position of the array in *O(1)*.

Insert function code for both implementations:

Adjacency matrix
```
char* insertX(int n, char arr[],
              char x, int pos)
{
    int i;
 
    // increase the size by 1
    n++;
 
    // shift elements forward
    for (i = n; i >= pos; i--)
        arr[i] = arr[i - 1];
 
    // insert x at pos
    arr[pos - 1] = x;
 
    return arr;
}
```

Adjacency list (not tested)
```
typedef struct line {
    char ch;
    struct node *next;
} line_t;

line_t *insertX(line_t *graph, char x, int pos)
{
    int i = 0;
    // to be inserted
    line_t *p;
    p = (line_t*)malloc(sizeof(line_t));
    p->ch = x;
    p->next = NULL;

    if (graph == NULL) {
        graph = p;
    } else {
        while (tmp->next != NULL) {
            tmp = tmp->next; //go forward
            i++;
            // ONLY this (below) part does the insertion
            if (i == pos) {
            //insert p to tmp
                p->next = tmp->next;
                tmp->next = p;
                break;
            }
        }
    }
    graph = tmp;

    return graph;
}
```
To sum up:
|--------------|----------------|------------------|
|              | Adjacency List | Adjacency Matrix |
| Random Access| O(n)           | O(1)             |
| Insertion    | O(1)           | O(n)             |
| Difficulty   | Difficult      | Easy             |

**Comment by ikozyris** [comment]
New issue found: Try inserting text in a line which has tabs '\t' not spaces.
A tab is a single character, yet it is displayed as 8 characters.
The current way of inserting, is inserting according to x-axis of the cursor.
So there is a 7 character offset.
